### PR TITLE
sys/net/gnrc/pkt: use `uint8_t` for user count

### PR DESCRIPTION
**Pull request by benpicco**
### Contribution description

It's very unlikely that a pkt snip will have more than 255 users. Use a `uint8_t` here to save 4 bytes per snip as this now fits into the unused struct padding. 

#### master

`sizeof(gnrc_pktsnip_t) == 20`

#### this PR

`sizeof(gnrc_pktsnip_t) == 12`

### CODE REVIEW

File: sys/net/gnrc/pktbuf_static/gnrc_pktbuf_static.c [hunk 1]
```diff
@@ -115,7 +115,7 @@ gnrc_pktsnip_t *gnrc_pktbuf_mark(gnrc_pktsnip_t *pkt, size_t size, gnrc_nettype_
     mutex_lock(&gnrc_pktbuf_mutex);
     if ((size == 0) || (pkt == NULL) || (size > pkt->size) || (pkt->data == NULL)) {
         DEBUG("pktbuf: size == 0 (was %" PRIuSIZE ") or pkt == NULL (was %p) or "
-              "size > pkt->size (was %" PRIuSIZE ") or pkt->data == NULL (was %p)\n",
```
Comment (maribu) [inline sys/net/gnrc/pktbuf_static/gnrc_pktbuf_static.c:118]: `%u` is for `unsigned int`, which may not be the same as `uint16_t`, e.g. on ARM. Yes, there is integer promotion, but that may also promote to `signed int`, which if that were guaranteed to be two's complement would be bitwise identical with the corresponding `unsigned int` for the value range, but the C standard chose to net mandate two's complement.

So yada yada please add a cast to `unsigned` in front of `pkt->size`.

Comment (maribu) [reply sys/net/gnrc/pktbuf_static/gnrc_pktbuf_static.c:118]: That's great news! 🎉

**Comment by miri64** [comment]
> We won't ever have pkt snips larger than 65535 bytes, so use `uint16_t` for the size - if it's good enough for UDP, it's good enough for us.

What about large payloads that later get fragmented?

**Review by miri64 — CHANGES_REQUESTED** [review]
I am fine with reducing the size of `users`, but I am not convinced that we only need `(1 << 16) - 1` of payload ever.

**Comment by miri64** [comment]
> We do not have drivers for hardware that can send frames larger than 1500 B.

There is fragmentation in both IP and IPv6 which are above the device hardware.

> Even if we would support [Jumbo frames](https://en.wikipedia.org/wiki/Jumbo_frame), those go typically up to at most 9000 B. And Jumbo frames are pretty exotic and error prone to set up.

If you want to talk Jumbo, the upper limit is `(1 << 32) - 1`: https://datatracker.ietf.org/doc/html/rfc2675 (so as it is right now).

> The only device that I could see actually sending large frames would be in SLIP. But given the bit error rate of UART, it is better to go with smaller frames.

Again, GNRC is a network stacks that has higher layers than link layer...

> And again as @benpicco said: We would need a custom transport protocol to be able to have more than 64 KiB of data, as UDP won't allow for more than 64 KiB.

Custom, exotic transport protocols such as TCP or QUIC? As the packet buffer is the main memory allocation arena for GNRC, we should not just think about a single packet payload, as it is sent on the wire.

**Comment by maribu** [comment]
> > Even if we would support [Jumbo frames](https://en.wikipedia.org/wiki/Jumbo_frame), those go typically up to at most 9000 B. And Jumbo frames are pretty exotic and error prone to set up.
> 
> If you want to talk Jumbo, the upper limit is `(1 << 32) - 1`: https://datatracker.ietf.org/doc/html/rfc2675 (so as it is right now).

That may be the theoretical limit. Hardware that does support Jumbo frames is already exotic. If they do, they typically do not get above 9K.

But honestly: What is the use case of packets larger than 64 KiB on MCUs?

**Comment by miri64** [comment]
> That may be the theoretical limit. Hardware that does support Jumbo frames is already exotic. If they do, they typically do not get above 9K.

Again: Think outside the hardware frame. A jumbogram would be a fragmented datagram, so the hardware will be totally fine when the network layer fragments the jumbogram in 9000 bytes packets.

> But honestly: What is the use case of packets larger than 64 KiB on MCUs?

Is RIOT a general purpose OS or not? We do support Ethernet after all. Let's not program ourselves into a corner that we might not get out of, just to optimize prematurely and because, _at least for now_, the limit is theoretical.

File: sys/include/net/gnrc/pkt.h [hunk 2]
```diff
@@ -109,18 +109,18 @@ typedef struct gnrc_pktsnip {
     /* the first three fields *MUST* match iolist_t! */
     struct gnrc_pktsnip *next;      /**< next snip in the packet */
     void *data;                     /**< pointer to the data of the snip */
-    size_t size;                    /**< the length of the snip in byte */
+    uint16_t size;                  /**< the length of the snip in byte */
```
Comment (miri64) [inline sys/include/net/gnrc/pkt.h:112]: Proposal for compromise: How about a typedef `gnrc_pkt_size_t`, which if need be can be increased in size with a simple MACRO switch?

Comment (benpicco) [reply sys/include/net/gnrc/pkt.h]: Ok

**Comment by miri64** [comment]
Would feel much better if someone ran those `tests/net/gnrc_tcp` and `tests/net/gnrc_sock*` tests on some real hardware.

File: sys/include/net/gnrc/pkt.h [hunk 3]
```diff
@@ -109,18 +119,18 @@ typedef struct gnrc_pktsnip {
     /* the first three fields *MUST* match iolist_t! */
     struct gnrc_pktsnip *next;      /**< next snip in the packet */
     void *data;                     /**< pointer to the data of the snip */
-    size_t size;                    /**< the length of the snip in byte */
+    gnrc_pkt_size_t size;           /**< the length of the snip in byte */
```
Comment (maribu) [inline sys/include/net/gnrc/pkt.h:122]: iolist_t needs to also be adjusted

Comment (benpicco) [reply sys/include/net/gnrc/pkt.h:122]: uh but why?

Comment (maribu) [reply sys/include/net/gnrc/pkt.h:122]: because it is casted to `iolist_t *` when passed to netdevs

Comment (benpicco) [reply sys/include/net/gnrc/pkt.h:122]: :fearful: 
then that's a can of worms that I rather leave the lid on

**Comment by kfessel** [comment]
a overflow check/assert in `gnrc_pktbuf_hold`
in 
sys/net/gnrc/pktbuf_static/gnrc_pktbuf_static.c
and 
sys/net/gnrc/pktbuf_static/gnrc_pktbuf_malloc.c

might be a good idea

**Comment by benpicco** [comment]
> `gnrc_pktbuf_hold` is sometime called with the interface number 

:exploding_head: 

but why? That comment is not very helpful

> `/* send packet to link layer */`

**Comment by kfessel** [comment]
> > `gnrc_pktbuf_hold` is sometime called with the interface number
> 
> 🤯
> 
> but why? That comment is not very helpful

I don't know i was just seaching the tree for some places that read or write to that value 
`gnrc_pktbuf_hold` seems to be the only place that is not just a 1 or  ++
 
When searching for uses of `gnrc_pktbuf_hold` i found strange things happening 

I also dont know why `gnrc_pktbuf_hold` would take a parameter (i would assume a count just goes up by one or down by one every time it is written), maybe that is some old api left over?

**Comment by kfessel** [comment]
> > `gnrc_pktbuf_hold` is sometime called with the interface number
> 
> 🤯
> 
> but why? That comment is not very helpful
> 
> > `/* send packet to link layer */`

checked again these are not interface numbers but number of interfaces to be called

both calls are ahead of loops that then release the pktbuf one by one

**Review by maribu — APPROVED** [review]

**Pull request merged**